File: linux/src/qmk_toolbox/bootloader/bootloader_factory.py
from .bootloader_device import BootloaderDevice
from .bootloader_type import BootloaderType
from ..usb.usb_device import UsbDevice


BOOTLOADER_VID_PID = {
    (0x03EB, 0x2FEF): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x2FF0): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x2FF3): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x2FF4): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x2FF9): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x2FFA): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x2FFB): BootloaderType.ATMEL_DFU,
    (0x03EB, 0x6124): BootloaderType.ATMEL_SAM_BA,
    (0x16C0, 0x0478): BootloaderType.HALFKAY,
    (0x16C0, 0x05DF): BootloaderType.BOOTLOAD_HID,
    (0x16C0, 0x05DC): BootloaderType.LUFA_HID,
    (0x1C11, 0xB007): BootloaderType.KIIBOHD_DFU,
    (0x1EAF, 0x0003): BootloaderType.STM32DUINO,
    (0x1209, 0x2301): BootloaderType.LUFA_MS,
    (0x0483, 0xDF11): BootloaderType.STM32_DFU,
    (0x314B, 0x0106): BootloaderType.APM32_DFU,
    (0x28E9, 0x0189): BootloaderType.GD32V_DFU,
    (0x342D, 0xDFA0): BootloaderType.WB32_DFU,
    (0x16C0, 0x0483): BootloaderType.AVR_ISP,
    (0x16C0, 0x05DC): BootloaderType.USBASP,
    (0x1781, 0x0C9F): BootloaderType.USBTINY_ISP,
    (0x2341, 0x0036): BootloaderType.CATERINA,
    (0x2341, 0x0037): BootloaderType.CATERINA,
    (0x2341, 0x8036): BootloaderType.CATERINA,
    (0x2341, 0x8037): BootloaderType.CATERINA,
    (0x1B4F, 0x9203): BootloaderType.CATERINA,
    (0x1B4F, 0x9205): BootloaderType.CATERINA,
    (0x1B4F, 0x9207): BootloaderType.CATERINA,
    (0x2A03, 0x0036): BootloaderType.CATERINA,
    (0x2A03, 0x0037): BootloaderType.CATERINA,
}


class BootloaderFactory:
    @staticmethod
    def create(usb_device: UsbDevice):
        key = (usb_device.vendor_id, usb_device.product_id)
        bootloader_type = BOOTLOADER_VID_PID.get(key)
        
        if not bootloader_type:
            return None
        
        if bootloader_type == BootloaderType.ATMEL_DFU:
            from .atmel_dfu_device import AtmelDfuDevice
            return AtmelDfuDevice(usb_device)
        elif bootloader_type == BootloaderType.STM32_DFU:
            from .stm32_dfu_device import Stm32DfuDevice
            return Stm32DfuDevice(usb_device)
        elif bootloader_type == BootloaderType.APM32_DFU:
            from .apm32_dfu_device import Apm32DfuDevice
            return Apm32DfuDevice(usb_device)
        elif bootloader_type == BootloaderType.KIIBOHD_DFU:
            from .kiibohd_dfu_device import KiibohdDfuDevice
            return KiibohdDfuDevice(usb_device)
        elif bootloader_type == BootloaderType.STM32DUINO:
            from .stm32duino_device import Stm32DuinoDevice
            return Stm32DuinoDevice(usb_device)
        elif bootloader_type == BootloaderType.GD32V_DFU:
            from .gd32v_dfu_device import Gd32vDfuDevice
            return Gd32vDfuDevice(usb_device)
        elif bootloader_type == BootloaderType.WB32_DFU:
            from .wb32_dfu_device import Wb32DfuDevice
            return Wb32DfuDevice(usb_device)
        elif bootloader_type == BootloaderType.CATERINA:
            from .caterina_device import CaterinaDevice
            return CaterinaDevice(usb_device)
        elif bootloader_type == BootloaderType.HALFKAY:
            from .halfkay_device import HalfKayDevice
            return HalfKayDevice(usb_device)
        elif bootloader_type == BootloaderType.BOOTLOAD_HID:
            from .bootload_hid_device import BootloadHidDevice
            return BootloadHidDevice(usb_device)
        elif bootloader_type == BootloaderType.LUFA_HID:
            from .lufa_hid_device import LufaHidDevice
            return LufaHidDevice(usb_device)
        elif bootloader_type == BootloaderType.LUFA_MS:
            from .lufa_ms_device import LufaMsDevice
            return LufaMsDevice(usb_device)
        elif bootloader_type == BootloaderType.ATMEL_SAM_BA:
            from .atmel_samba_device import AtmelSamBaDevice
            return AtmelSamBaDevice(usb_device)
        elif bootloader_type in (BootloaderType.AVR_ISP, BootloaderType.USBASP, BootloaderType.USBTINY_ISP):
            from .isp_device import IspDevice
            return IspDevice(usb_device, bootloader_type)
        
        return None

File: src/QmkToolbox.Core/Bootloader/BootloaderFactory.cs
using QmkToolbox.Core.Bootloader.Impl;
using QmkToolbox.Core.Models;
using QmkToolbox.Core.Services;

namespace QmkToolbox.Core.Bootloader;

public static class BootloaderFactory
{
    private static readonly Dictionary<(ushort VID, ushort PID), BootloaderType> DeviceMap = new()
    {
        // ── Atmel Corporation (0x03EB) ──────────────────────────────────
        [(0x03EB, 0x2045)] = BootloaderType.LufaMs,
        [(0x03EB, 0x2067)] = BootloaderType.LufaHid,
        [(0x03EB, 0x2FEF)] = BootloaderType.AtmelDfu, // ATmega16U2
        [(0x03EB, 0x2FF0)] = BootloaderType.AtmelDfu, // ATmega32U2
        [(0x03EB, 0x2FF3)] = BootloaderType.AtmelDfu, // ATmega16U4
        [(0x03EB, 0x2FF4)] = BootloaderType.AtmelDfu, // ATmega32U4
        [(0x03EB, 0x2FF9)] = BootloaderType.AtmelDfu, // AT90USB64
        [(0x03EB, 0x2FFA)] = BootloaderType.AtmelDfu, // AT90USB162
        [(0x03EB, 0x2FFB)] = BootloaderType.AtmelDfu, // AT90USB128
        [(0x03EB, 0x6124)] = BootloaderType.AtmelSamBa,

        // ── STMicroelectronics (0x0483) ─────────────────────────────────
        [(0x0483, 0xDF11)] = BootloaderType.Stm32Dfu,

        // ── pid.codes (0x1209) ──────────────────────────────────────────
        [(0x1209, 0x2302)] = BootloaderType.Caterina,

        // ── Van Ooijen Technische Informatica (0x16C0) ──────────────────
        [(0x16C0, 0x0478)] = BootloaderType.HalfKay,
        [(0x16C0, 0x0483)] = BootloaderType.AvrIsp,
        [(0x16C0, 0x05DC)] = BootloaderType.UsbAsp,
        [(0x16C0, 0x05DF)] = BootloaderType.BootloadHid,

        // ── MECANIQUE (0x1781) ──────────────────────────────────────────
        [(0x1781, 0x0C9F)] = BootloaderType.UsbTinyIsp,

        // ── Spark Fun Electronics (0x1B4F) ──────────────────────────────
        [(0x1B4F, 0x9203)] = BootloaderType.Caterina,
        [(0x1B4F, 0x9205)] = BootloaderType.Caterina,
        [(0x1B4F, 0x9207)] = BootloaderType.Caterina,

        // ── Input Club Inc. (0x1C11) ────────────────────────────────────
        [(0x1C11, 0xB007)] = BootloaderType.KiibohdDfu,

        // ── Leaflabs (0x1EAF) ───────────────────────────────────────────
        [(0x1EAF, 0x0003)] = BootloaderType.Stm32Duino,

        // ── Pololu Corporation (0x1FFB) ─────────────────────────────────
        [(0x1FFB, 0x0101)] = BootloaderType.Caterina,

        // ── Arduino SA (0x2341) ─────────────────────────────────────────
        [(0x2341, 0x0036)] = BootloaderType.Caterina,
        [(0x2341, 0x0037)] = BootloaderType.Caterina,

        // ── Adafruit (0x239A) ───────────────────────────────────────────
        [(0x239A, 0x000C)] = BootloaderType.Caterina,
        [(0x239A, 0x000D)] = BootloaderType.Caterina,
        [(0x239A, 0x000E)] = BootloaderType.Caterina,

        // ── dog hunter AG (0x2A03) ──────────────────────────────────────
        [(0x2A03, 0x0036)] = BootloaderType.Caterina,
        [(0x2A03, 0x0037)] = BootloaderType.Caterina,
        [(0x2A03, 0x0040)] = BootloaderType.Caterina,

        // ── GigaDevice Semiconductor (0x28E9) ──────────────────────────
        [(0x28E9, 0x0189)] = BootloaderType.Gd32VDfu,

        // ── ArteryTek (0x2E3C) ──────────────────────────────────────────
        [(0x2E3C, 0xDF11)] = BootloaderType.At32Dfu,

        // ── Geehy Semiconductor (0x314B) ────────────────────────────────
        [(0x314B, 0x0106)] = BootloaderType.Apm32Dfu,

        // ── WestBerryTech (0x342D) ──────────────────────────────────────
        [(0x342D, 0xDFA0)] = BootloaderType.Wb32Dfu,

        // ── Raspberry Pi (0x2E8A) ────────────────────────────────────────
        [(0x2E8A, 0x0003)] = BootloaderType.Picotool, // RP2040 BOOTSEL
        [(0x2E8A, 0x000F)] = BootloaderType.Picotool, // RP2350 BOOTSEL
    };

    public static BootloaderDevice? CreateDevice(
        IUsbDevice device,
        IFlashToolProvider toolProvider,
        ISerialPortService? serialPortService = null,
        IMountPointService? mountPointService = null)
    {
        return GetDeviceType(device.VendorId, device.ProductId, device.RevisionBcd) switch
        {
            BootloaderType.Apm32Dfu => new Apm32DfuDevice(device, toolProvider),
            BootloaderType.At32Dfu => new At32DfuDevice(device, toolProvider),
            BootloaderType.AtmelDfu => new AtmelDfuDevice(device, toolProvider, serialPortService),
            BootloaderType.AtmelSamBa => new AtmelSamBaDevice(device, toolProvider, serialPortService),
            BootloaderType.AvrIsp => new AvrIspDevice(device, toolProvider, serialPortService),
            BootloaderType.BootloadHid => new BootloadHidDevice(device, toolProvider),
            BootloaderType.Caterina => new CaterinaDevice(device, toolProvider, serialPortService),
            BootloaderType.Gd32VDfu => new Gd32VDfuDevice(device, toolProvider),
            BootloaderType.HalfKay => new HalfKayDevice(device, toolProvider),
            BootloaderType.KiibohdDfu => new KiibohdDfuDevice(device, toolProvider),
            BootloaderType.LufaHid => new LufaHidDevice(device, toolProvider),
            BootloaderType.LufaMs => new LufaMsDevice(device, toolProvider, mountPointService),
            BootloaderType.Stm32Dfu => new Stm32DfuDevice(device, toolProvider),
            BootloaderType.Stm32Duino => new Stm32DuinoDevice(device, toolProvider),
            BootloaderType.UsbAsp => new UsbAspDevice(device, toolProvider),
            BootloaderType.UsbTinyIsp => new UsbTinyIspDevice(device, toolProvider),
            BootloaderType.Wb32Dfu => new Wb32DfuDevice(device, toolProvider),
            BootloaderType.Picotool => new PicotoolDevice(device, toolProvider),
            BootloaderType.QmkDfu => new AtmelDfuDevice(device, toolProvider, serialPortService),
            BootloaderType.QmkHid => new LufaHidDevice(device, toolProvider),
            BootloaderType.None => null,
            var type => throw new ArgumentOutOfRangeException(nameof(BootloaderType), type, "No device implementation for this bootloader type"),
        };
    }

    public static BootloaderType GetDeviceType(ushort vendorId, ushort productId, ushort revisionBcd)
    {
        if (!DeviceMap.TryGetValue((vendorId, productId), out BootloaderType type))
            return BootloaderType.None;
        // The QMK revision marker distinguishes QMK-flavoured bootloaders sharing stock VID/PIDs.
        if (revisionBcd == BootloaderDevice.QmkRevisionMarker)
        {
            if (type == BootloaderType.AtmelDfu)
                return BootloaderType.QmkDfu;
            if (type == BootloaderType.LufaHid)
                return BootloaderType.QmkHid;
        }
        return type;
    }
}
